File: IAIso-v5.0/core/iaiso-ruby/lib/iaiso/middleware/mistral.rb
# frozen_string_literal: true

require_relative "errors"
require_relative "../core/pressure_engine"
require_relative "../core/step_outcome"

module IAIso
  module Middleware
    module Mistral
      Usage = Data.define(:prompt_tokens, :completion_tokens, :total_tokens) do
        def self.empty = new(prompt_tokens: 0, completion_tokens: 0, total_tokens: 0)
      end
      Choice = Data.define(:tool_calls) do
        def self.empty = new(tool_calls: [])
      end
      Response = Data.define(:model, :usage, :choices)

      Options = Data.define(:raise_on_escalation) do
        def self.defaults = new(raise_on_escalation: false)
      end

      class BoundedClient
        def initialize(raw:, execution:, options: Options.defaults)
          @raw = raw
          @execution = execution
          @options = options
        end

        def chat_complete(params)
          pre = @execution.check
          raise LockedError if pre == IAIso::Core::StepOutcome::LOCKED
          if pre == IAIso::Core::StepOutcome::ESCALATED && @options.raise_on_escalation
            raise EscalationRaisedError
          end
          resp =
            begin
              @raw.chat_complete(params)
            rescue StandardError => e
              raise ProviderError, e.message
            end
          tokens = resp.usage.total_tokens.to_i
          tokens = resp.usage.prompt_tokens.to_i + resp.usage.completion_tokens.to_i if tokens.zero?
          tool_calls = resp.choices.sum { |c| c.tool_calls.size }
          model = resp.model.to_s.empty? ? "unknown" : resp.model
          @execution.record_step(IAIso::Core::StepInput.build(
            tokens: tokens, tool_calls: tool_calls,
            tag: "mistral.chat.complete:#{model}",
          ))
          resp
        end
      end
    end
  end
end

File: IAIso-v5.0/core/iaiso-csharp/src/Iaiso.Middleware/Mistral.cs
using System.Collections.Generic;
using System.Text.Json.Nodes;
using Iaiso.Core;

namespace Iaiso.Middleware.Mistral;

/// <summary>IAIso wrapper for Mistral chat.</summary>
public interface IMistralClient
{
    MistralResponse ChatComplete(JsonObject @params);
}

public sealed class MistralUsage
{
    public long PromptTokens { get; set; }
    public long CompletionTokens { get; set; }
    public long TotalTokens { get; set; }
}

public sealed class MistralChoice
{
    public IReadOnlyList<string> ToolCalls { get; }
    public MistralChoice(IReadOnlyList<string>? toolCalls)
    {
        ToolCalls = toolCalls ?? new List<string>();
    }
}

public sealed class MistralResponse
{
    public string Model { get; }
    public MistralUsage Usage { get; }
    public IReadOnlyList<MistralChoice> Choices { get; }
    public MistralResponse(string model, MistralUsage? usage, IReadOnlyList<MistralChoice>? choices)
    {
        Model = model;
        Usage = usage ?? new MistralUsage();
        Choices = choices ?? new List<MistralChoice>();
    }
}

public sealed class MistralOptions
{
    public bool RaiseOnEscalation { get; set; }
    public static MistralOptions Defaults() => new();
}

public sealed class BoundedMistralClient
{
    private readonly IMistralClient _raw;
    private readonly BoundedExecution _execution;
    private readonly MistralOptions _opts;

    public BoundedMistralClient(IMistralClient raw, BoundedExecution execution, MistralOptions? opts = null)
    {
        _raw = raw;
        _execution = execution;
        _opts = opts ?? MistralOptions.Defaults();
    }

    public MistralResponse ChatComplete(JsonObject @params)
    {
        var pre = _execution.Check();
        if (pre == StepOutcome.Locked) throw new MiddlewareException.Locked();
        if (pre == StepOutcome.Escalated && _opts.RaiseOnEscalation)
            throw new MiddlewareException.EscalationRaised();

        MistralResponse resp;
        try { resp = _raw.ChatComplete(@params); }
        catch (System.Exception e) { throw new MiddlewareException.Provider(e.Message, e); }

        long tokens = resp.Usage.TotalTokens;
        if (tokens == 0) tokens = resp.Usage.PromptTokens + resp.Usage.CompletionTokens;
        long toolCalls = 0;
        foreach (var c in resp.Choices) toolCalls += c.ToolCalls.Count;
        string model = string.IsNullOrEmpty(resp.Model) ? "unknown" : resp.Model;
        _execution.RecordStep(new StepInput
        {
            Tokens = tokens,
            ToolCalls = toolCalls,
            Tag = $"mistral.chat.complete:{model}",
        });
        return resp;
    }
}
